cpp/examples/tutorial_examples/dataset_example.cc

```diff
@@ -0,0 +1,177 @@
+  // from. Don't worry, we also write a dataset in the example proper.
+  auto base_path = root_path + "parquet_dataset";
+  ARROW_RETURN_NOT_OK(filesystem->CreateDir(base_path));
+  // Create an Arrow Table
+  ARROW_ASSIGN_OR_RAISE(auto table, CreateTable());
+  // Write it into two Parquet files
+  ARROW_ASSIGN_OR_RAISE(auto output,
+                        filesystem->OpenOutputStream(base_path + "/data1.parquet"));
+  ARROW_RETURN_NOT_OK(parquet::arrow::WriteTable(
+      *table->Slice(0, 5), arrow::default_memory_pool(), output, 2048));
+  ARROW_ASSIGN_OR_RAISE(output,
+                        filesystem->OpenOutputStream(base_path + "/data2.parquet"));
+  ARROW_RETURN_NOT_OK(parquet::arrow::WriteTable(
+      *table->Slice(5), arrow::default_memory_pool(), output, 2048));
+  return base_path;
+}
+
+arrow::Status PrepareEnv() {
+  // Get our environment prepared for reading, by setting up some quick writing.
+  ARROW_ASSIGN_OR_RAISE(auto src_table, CreateTable())
+  std::shared_ptr<arrow::fs::FileSystem> setup_fs;
+  // Note this operates in the directory the executable is built in.
+  char setup_path[256];
+  getcwd(setup_path, 256);
+  ARROW_ASSIGN_OR_RAISE(setup_fs, arrow::fs::FileSystemFromUriOrPath(setup_path));
+  ARROW_ASSIGN_OR_RAISE(auto dset_path, CreateExampleParquetDataset(setup_fs, ""));
+
+  return arrow::Status::OK();
+}
+
+arrow::Status RunMain() {
+  ARROW_RETURN_NOT_OK(PrepareEnv());
+
+  // First, we need a filesystem object, which lets us interact with our local
+  // filesystem starting at a given path. For the sake of simplicity, that'll be
+  // the current directory.
+  std::shared_ptr<arrow::fs::FileSystem> fs;
+  // This feels pretty bad, but I wasn't finding great solutions that're
+  // system-generic -- could use some advice on how to set this up.
+  char init_path[256];
+  getcwd(init_path, 256);
+  ARROW_ASSIGN_OR_RAISE(fs, arrow::fs::FileSystemFromUriOrPath(init_path));
+
+  // A file selector lets us actually traverse a multi-file dataset.
+  arrow::fs::FileSelector selector;
+  selector.base_dir = "parquet_dataset";
+  // Recursive is a safe bet if you don't know the nesting of your dataset.
+  selector.recursive = true;
+  // Making an options object lets us configure our dataset reading.
+  arrow::dataset::FileSystemFactoryOptions options;
+  // We'll use Hive-style partitioning. We'll let Arrow Datasets infer the partition
+  // schema. We won't set any other options, defaults are fine.
+  options.partitioning = arrow::dataset::HivePartitioning::MakeFactory();
+  auto read_format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+  // Now, we get a factory that will let us get our dataset -- we don't have the
+  // dataset yet!
+  ARROW_ASSIGN_OR_RAISE(auto factory, arrow::dataset::FileSystemDatasetFactory::Make(
+                                          fs, selector, read_format, options));
+  // Now we read into our dataset from the factory.
```

Comment:
The wording is a bit odd to me here since there isn't any data being read - the factory is a factory/builder in Java/OOP parlance

Reply:
What would be the proper wording here? I'm happy to change it from "read."

Reply:
Construct, build, or make I suppose